### ISSUE TITLE
Investigate usage of RedHat UBI

### ISSUE THREAD

**Issue by michaelsauter** [opendevstack/ods-quickstarters#7]
Figure out if https://www.redhat.com/en/blog/introducing-red-hat-universal-base-image would be a better default choice than alpine.

**Comment by renedupont** [comment]
@michaelsauter any suggestions what characteristics UBI should have to be favored and how we could/should test these?

The only thing I can come up with so far is that UBI might be stronger regarding enterprise-grade security capabilities and better maintained.

But looking at size, I pulled both and there seems to be a huge diff:

REPOSITORY                                TAG                 IMAGE ID            CREATED             SIZE
registry.access.redhat.com/ubi8-minimal   latest              a5d8ad363d87        2 weeks ago         142MB
alpine                                    latest              a24bb4013296        2 weeks ago         5.57MB

**Comment by michaelsauter** [comment]
@renedupont Oh wow, I expected it to be less than 142MB.

Anyway, I think UBIs make sense once we go to OpenShift 4. There, the Jenkins master and agent images are UBI based. If also the runtime images are UBI based, then we have the same architecture/OS in both containers, which avoids some nasty problems with native extensions (e.g. when using Python, JavaScript etc.)

Basically I would say right now we can use either UBI or Alpine, but once our agent images are UBI based, it makes sense to align.

**Comment by renedupont** [comment]
Ok sounds good @michaelsauter . Do you see any reason to keep this ticket open then?

**Comment by michaelsauter** [comment]
Yes, I think it's good to keep it there as the future should bring that change eventually.

**Comment by clemensutschig** [comment]
@michaelsauter  . this is done, or ... so I guess we can close this?

**Comment by michaelsauter** [comment]
This is not done yet. The base images of the quickstarters are still Alpine based. We recently touched this topic during a discussion on patch management (@metmajer @maksymmaksym). We agreed that it would be good to explore using RedHat provided base images from a security perspective. From a developer perspective, it's also attractive because having the same architecture in the image we're building with as in the runtime image avoids confusion / mistakes / doing work twice. On the other hand, UBI image are quite limited so we might have people hitting those limits.

@metmajer Is this something you'd want to have in ODS 4?

**Comment by metmajer** [comment]
@michaelsauter I would not rush here. In a recent meeting with @maksymmaksym we discussed that this will not be part of the next EDP 1.2/ODS 4 release.

**Comment by michaelsauter** [comment]
Sounds good to me. Moving off of ODS 4.0.0.

**Comment by michaelsauter** [comment]
@metmajer What alternative to moving to UBI do you see? Pulling from DockerHub directly blocks a lot of people who all have to do their workaround. I think it would be good if the next release has some solution to it - or some other non-release related remedy.

**Comment by metmajer** [comment]
@michaelsauter maybe a misunderstanding? I am fully with you regarding UBI, but have to mention that we (my teams) won't have the ability to spend time on this topic until the next release. I mean to remember that @gerardcl mentioned in this thread (maybe changed in the meantime?) that UBI could be 1 of several options, and interpreted this as: maintain multiple base images in parallel which I am not in favour of.

**Comment by gerardcl** [comment]
hi @metmajer , I have never proposed to have different base images / options for each quickstarter, rather to tackle each quickstarter the better way we find. For the Python-related ones, I found we can already use the UBI8 images, which is already in place in master branch -> ready for ODS4. 

Regarding other quickstarters, we might, for example, end up having to provide a caching-like solution. For example, for the be-gateway-nginx we could either have a template that already comes with an imagestream that caches the docker.io image or we could also have that provided within the ods (ods-core) namespace registry (since all projects have access)...

**Comment by metmajer** [comment]
@gerardcl makes sense! I must have misinterpreted something when skimming over an email recently. Seems we're aligned and, granted, UBI images won't be the best solution everywhere!

**Comment by segfault16** [comment]
Regarding Docker Rate Limits on Dockerhub:

This problem can occur in

- Openshift Docker Image Build (when using external image as base)
- Scheduling of Jenkins Agent Pods if a podTemplate with an external image is used
- Scheduling of Application Pods if external image is used, e.g. as a sidecar

Proposed fixes so far:

- Transparent proxy for docker images configured correctly throughout Openshift (e.g. with Nexus)
    - (+) Decoupling from the internet, reduction of network traffic
    -  (+) Transparent to developers
    - (+) IMHO the cleanest solution used throughout the industry
- "Caching" of Docker images by using only "FROM image:tag" and pushing into Openshift Docker Registry
    - (+) Viable solution for usage in Jenkins Agent Pods
    - (-) Not applicable (at least for me) in Application Pods, since I need to be able to build docker images locally for certain debugging tasks. It's one key feature why we use docker images in the first place: reproducability. With the internal docker registry unaccessible from my dev machines this produces more problems than it solves
    - (-) Potential risk of using outdated images (e.g. without important CVE's fixed on base image layer)
- Use base images from other public docker registry		
    - (-) Need to re-do all our Dockerfiles
    - (-) Need to apply other solutions to stuff that I can only use from Dockerhub
    - (-) (Need to configure Pull Secrets depending on which other registry used?)
- Provide image pull secrets for Dockerhub
    - (-) Not applicable in Openshift Image Build
    - (-) Needs to be done by each project, legal evaluation? cost?

**Comment by metmajer** [comment]
@gerardcl @michaelsauter do we know already where UBI images are not a good fit? I assume that the quickstarter owners handle the migration to UBI where needed? Please consider that, now with the use of Aqua, vulnerabilities will be alerted in production, which emphasises the use of secure base images.

FYI @maksymmaksym

**Comment by michaelsauter** [comment]
At least I haven't done much research so far. However, I had some touchpoints with UBI images in the https://github.com/opendevstacl/ods-pipeline project. There we use only UBI based images, and this works well. We are able to build everything locally (pulling unauthenticated from access.registry.redhat.com) and in the cluster (pulling from registry.redhat.io). We already have images to build Go, Java, Python and TypeScript which worked just fine. Runtime images are obviously not part of ods-pipeline but typically these are far simpler (e.g. in case of Go you do not even need the toolchain). Therefore I would say at least for the main technologies it looks very promising.

To @segfault16's points, the UBI approach would fall under "Use base images from other public docker registry" I believe and while I do agree with your downsides listed there, UBI images have a huge upside as well in terms of vulnerability management, which I believe would outweigh the downsides from a platform perspective.

**Comment by metmajer** [comment]
@michaelsauter I see it the same. I am also scared about an unpredictable number of incidents in production due to non-maintained base images. Shall we plan a slim release end of summer on top of 4.x that migrates to UBI images?

**Comment by michaelsauter** [comment]
@metmajer Makes sense to me. I think the biggest effort is not updating the GitHub repo, but pushing that change through the organisation as every repo needs to be touched to change the `Dockerfile`.

**Comment by metmajer** [comment]
@michaelsauter we are running a migration project for all use-cases on OCP3 -> OCP4. The sooner the better. If it's simple, let's do it now, as part of ODS 4.0.

**Comment by gerardcl** [comment]
hi! what are the expected timelines? 

for me it also make sense, keeping in mind that users might change that anyway, but that is a known and expected behaviour which then should be tackled at team level (vulnerabilities found and so on...) 

-> we could run a small spike with all the related responsibles of the quickstarters and make this happen? I guess some of us are now on vacations but maybe by end of August we could jump on it?

**Comment by maksymmaksym** [comment]
It will be very inefficient if every project in prod cluster would need to fix the same vulns over and over
